### PR TITLE
[SPARK-8300] DataFrame hint for broadcast join.

### PR DESCRIPTION
**Pull request by rxin**
Users can now do

``` scala
left.join(broadcast(right), "joinKey")
```

to give the query planner a hint that "right" DataFrame is small and should be broadcasted.

### CODE REVIEW

**Comment by patmcdonough** [comment]
@rxin - This is great. I'm a fan of the alternative syntax we chatted about:

~~~
left.join(right.broadcast, "joinKey")
~~~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 1]
```diff
@@ -130,6 +130,14 @@ case class Join(
   }
 }
 
+/**
+ * A hint for the optimizer that we should broadcast the `child` if used in a join operator.
+ */
+case class BroadcastHint(child: LogicalPlan) extends UnaryNode {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Could we just have this be `StatisticsHint` and override the statistics?  I'm afraid that we are going to forget to add cases in the future as we broadcast more type of joins.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: that sounds good.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: I'm thinking about this more and actually maybe what we want to do is have this specific node, and a patter that recognizes `canBroadcast`.  This pattern can check either for a small enough size or this hint and we can use that anywhere we are planning a broadcast operator.  The reasoning is that messing with statistics could have other consequences.

**Comment by marmbrus** [comment]
Regarding `broadcast(...)` vs `.broadcast`, I don't really care too much.  However, I think we need to come up with a convention.  For example, I was recently confused that its not `isNotNull(column)` but instead `column.isNotNull`.

**Comment by rxin** [comment]
We should add isNotNull too.

The isNotNull was from SchemaRDD.

**Comment by marmbrus** [comment]
I'm confused now: both `column.isNotNull` and `column.isNull` work today.  I was expecting them to be functions instead of being functions that column defines.  I'm mostly wondering how we decide what goes where.

**Comment by rxin** [comment]
I meant adding isNotNull as a function, rather than a member method of Column.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 1 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: +1
Add specific node is also doable, probably the node can be named like `Hint` in `LogicalPlan`, and could be more than `broadcast`, like `uniq_key` etc. 

Any idea?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -53,6 +53,18 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   }
 
   /**
+   * Matches a plan whose output should be small enough to be used in broadcast join.
+   */
+  object CanBroadcast {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @marmbrus is this what you had in mind?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Yep!

**Comment by marmbrus** [comment]
Merging to master!

**Comment by sridharsubramanian62** [comment]
How can this hint be used with Spark-SQL ?

Below example does not seem to work. 

```select a.col1,b.col2 from tableA a left outer join broadcast(tableB) on a.col1 = b.col1```

**Comment by marmbrus** [comment]
It can only be used from the dataframe API.

**Comment by RoiViber** [comment]
Using broadcast(...) results in "java.lang.AssertionError: assertion failed: No plan for BroadcastHint" as in https://issues.apache.org/jira/browse/SPARK-12275 so this solution is not working most of the time

**Comment by fjh100456** [comment]
@rxin  @marmbrus 
Is there another way to broadcast table with the spark-sql now, except by `spark.sql.autoBroadcastJoinThreshold`?
And if no, is it a good way to broadcast table by user conf,such as `spark.sql.autoBroadcastJoin.mytable=true` to broadcast a table named `mytable`.

**Comment by rxin** [comment]
Isn't the hint available in SQL?

**Comment by sridharsubramanian62** [comment]
Its available from spark 2.2.0.
    On Tuesday, October 10, 2017, 1:46:27 PM PDT, Reynold Xin <notifications@github.com> wrote:  
 
 
Isn't the hint available in SQL?

**Comment by fjh100456** [comment]
Is there an example?
I use broadcast like the following, but it perform an error.Would you be so kind as to show me an example?

>spark-sql> select a.* from tableA a left outer join broadcast(tableB) b on a.a=b.a;
>17/10/11 09:06:40 INFO HiveMetaStore: 0: get_table : db=default tbl=tablea
>17/10/11 09:06:40 INFO audit: ugi=root	ip=unknown-ip-addr	cmd=get_table : db=default tbl=tablea	
> #### **Error in query: cannot resolve '`tableB`' given input columns: []; line 1 pos 51;**
>'Project [ArrayBuffer(a).*]
>+- 'Join LeftOuter, ('a.a = 'b.a)
>   :- SubqueryAlias a
>   :  +- SubqueryAlias tablea
>   :     +- HiveTableRelation `default`.`tablea`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [a#94, b#95, c#96]
>   +- 'SubqueryAlias b
>      +- 'UnresolvedTableValuedFunction broadcast, ['tableB]

**Comment by sridharsubramanian62** [comment]
Let me know if this does not help
[SPARK-16475] Broadcast Hint for SQL Queries - ASF JIRA


| 
| 
|  | 
[SPARK-16475] Broadcast Hint for SQL Queries - ASF JIRA


 |

 |

 |




    On Tuesday, October 10, 2017, 6:25:57 PM PDT, fjh100456 <notifications@github.com> wrote:  
 
 
Is there an example?
I use broadcast like the following, but it perform an error.Would you be so kind as to show me an example?


spark-sql> select a.* from tableA a left outer join broadcast(tableB) b on a.a=b.a;
17/10/11 09:06:40 INFO HiveMetaStore: 0: get_table : db=default tbl=tablea
17/10/11 09:06:40 INFO audit: ugi=root ip=unknown-ip-addr cmd=get_table : db=default tbl=tablea

Error in query: cannot resolve 'tableB' given input columns: []; line 1 pos 51;

'Project [ArrayBuffer(a).*]
+- 'Join LeftOuter, ('a.a = 'b.a)
:- SubqueryAlias a
: +- SubqueryAlias tablea
: +- HiveTableRelation default.tablea, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [a#94, b#95, c#96]
+- 'SubqueryAlias b
+- 'UnresolvedTableValuedFunction broadcast, ['tableB]

**Comment by fjh100456** [comment]
With `/*+ broadcast(table) */`, it works well, thank you very much.

**Pull request closed without merging**